contracts/src/subnet/SubnetActorCheckpointingFacet.sol

```diff
@@ -96,17 +89,30 @@ contract SubnetActorCheckpointingFacet is SubnetActorModifiers, ReentrancyGuard,
             revert MaxMsgsPerBatchExceeded();
         }
 
-        // if the bottom up messages' length is max, we consider that epoch valid
-        if (checkpoint.msgs.length == s.maxMsgsPerBottomUpBatch) {
+        uint256 lastBottomUpCheckpointHeight = s.lastBottomUpCheckpointHeight;
+        uint256 bottomUpCheckPeriod = s.bottomUpCheckPeriod;
+
+        // cannot submit past bottom up checkpoint
+        if (checkpoint.blockHeight <= lastBottomUpCheckpointHeight) {
+            revert BottomUpCheckpointAlreadySubmitted();
+        }
+
+        uint256 nextCheckpointHeight = LibGateway.getNextEpoch(lastBottomUpCheckpointHeight, bottomUpCheckPeriod);
+
+        if (checkpoint.blockHeight > nextCheckpointHeight) {
+            revert CannotSubmitFutureCheckpoint();
+        }
+
+        // the expected bottom up checkpoint height, valid height
+        if (checkpoint.blockHeight == nextCheckpointHeight) {
```

Comment:
This does not line up with my understanding Fendermint, which is also not being updated in this PR. So I think this will PR will break checkpointing.

https://github.com/consensus-shipyard/ipc/blob/9509c75c3887f280a4e521ac54d549a51ffbe06d/fendermint/vm/interpreter/src/fvm/checkpoint.rs#L328

I believe Fendermint produces checkpoints every `bottomUpCheckPeriod` from genesis. If an early checkpoint was produced due to reaching the msg batch limit, Fendermint will still produce the next checkpoint at the original schedule, but this contract change will expect it at _exactly_ `bottomUpCheckPeriod` epochs after the last checkpoint.

---

Test screenshot you shared with me also confirms my understanding is correct:

![image](https://github.com/consensus-shipyard/ipc/assets/1101242/47344a22-95c9-4cf0-8443-004df2b21c5f)

Did you verify that the checkpoint at 1500 was accepted by the subnet actor at the parent?

Reply:
I think the confusion comes from this line:
```
uint256 nextCheckpointHeight = LibGateway.getNextEpoch(lastBottomUpCheckpointHeight, bottomUpCheckPeriod);
```
This function basically deduces the next checkpoint height, see [test cases](https://github.com/consensus-shipyard/ipc/pull/824/files#diff-b5777dbc3124d3a87f7b1eb83ab1f96306a662b3fd0fd93c0f27a7c9507c50ccR430) added.

Reply:
Logic seems correct, apologies for the false alarm! I'd suggest improving the naming of the method at least (`LibGateway.getNextEpoch()` is missing information).